mayastor/src/bdev/nexus/nexus_io.rs

```diff
@@ -1,471 +1,478 @@
+                IoStatus::Pending => Disposition::Flying(IoStatus::Success),
+                IoStatus::Failed => Disposition::Flying(IoStatus::Failed),
+                IoStatus::NoMemory => Disposition::Flying(IoStatus::NoMemory),
+                _ => panic!("should not happen"),
             }
         }
     }
 
-    /// assess the IO if we need to mark it failed or ok.
-    #[inline]
-    pub(crate) fn assess(&mut self, child_io: &mut Bio, success: bool) {
-        self.ctx_as_mut_ref().dec();
+    /// returns the type of command for this IO
+    #[inline(always)]
+    fn cmd(&self) -> IoType {
+        self.io_type()
+    }
 
+    /// completion handler for the nexus when a child IO completes
+    pub fn complete(&mut self, child_io: Bio, success: bool) {
+        // decrement the counter of in flight IO
+        self.ctx_as_mut().in_flight -= 1;
+        assert_eq!(self.ctx().core, Cores::current());
+        // trace!("{:p} - {:p}", self, child_io);
+        // record the state of at least one of the IO's.
         if !success {
-            // currently, only tests send those but invalid op codes should not
-            // result into faulting a child device.
-            if NvmeStatus::from(child_io.clone()).status_code()
-                == GenericStatusCode::InvalidOpcode
-            {
-                self.complete();
-                return;
-            }
-
-            // all other status codes indicate a fatal error
-            Reactors::master().send_future(Self::child_retire(
-                self.nexus_as_ref().name.clone(),
-                child_io.bdev_as_ref(),
-            ));
+            self.ctx_as_mut().status = IoStatus::Failed;
         }
 
-        self.complete();
-    }
-
-    async fn child_retire(nexus: String, child: Bdev) {
-        error!("{:#?}", child);
-
-        if let Some(nexus) = nexus_lookup(&nexus) {
-            if let Some(child) = nexus.child_lookup(&child.name()) {
-                let current_state = child.state.compare_and_swap(
-                    ChildState::Open,
-                    ChildState::Faulted(Reason::IoError),
+        // determine what to do now based on the context specific info of the IO
+        match self.disposition() {
+            // the happy path, all is good
+            Disposition::Complete(IoStatus::Success) => self.ok(),
+            // All of IO's have failed but all remaining in flights completed now
+            // as well depending on the error we can attempt to do a
+            // retry.
+            Disposition::Complete(IoStatus::Failed) => self.fail(),
+
+            // IOs were submitted before we bumped into an error. The IO has
+            // now completed, so we can finally report back to the
+            // callee that we encountered ENOMEM while processing
+            Disposition::Complete(IoStatus::NoMemory) => self.no_mem(),
+
+            // remove child if any from the nexus
+            Disposition::Flying(IoStatus::Failed) => {
```

Comment:
Do we consider a situation when I/O error occurs exactly on the last replica involved in Nexus I/O ? Let's assume a situation with 3 replicas, so we need to distribute our I/O across these 3 replicas. We can observe the following sequence of events (initial in_flight value is 3):

1. First replica completes successfully,  in_flight = 2, disposition = Disposition::Flying(IoStatus::Success)
2. Second replica completes successfully,  in_flight = 1, disposition = Disposition::Flying(IoStatus::Success)
3. Third replica ,  in_flight = 0, ctx.status = IoStatus::Failed, disposition = Disposition::Complete(IoStatus::Failed)

Shall we retire the third replica too, because currently child removal is done only in response to Disposition::Flying(IoStatus::Failed), which means all replicas but not the last serving the I/O.

Reply:
Good question. Im not sure. Currently, the behaviour is such that we always retire the IO, and thus have a target, children. We then keep the controller in the paused state which is only valid for the duration of restarts/reboots of the FW. I don't think it's OK to leave it in that state. With this approach, we always keep the target in a valid state all IO's will error out accordingly.

Put differently, we retire up to N-1 children of an N-way mirror which to me seems the right thing to do for now.

Reply:
Such a situation is not related to the number of replicas but rather to the sequence of failures, observed on replicas. I mean, the following 2 sequences of I/O completions (among 3 replicas) will be handled differently, although they seem to be identical to me:

1. Succces, Failure, Success  => Replica 2 will be retired.
2. Success, Success, Failure => Replica 3 will not be retired.

Reply:
Right ok got it, I misunderstood your question. That indeed is exactly what happens right now and should be handled as well :+1:

Reply:
@mtzaurus I added the case you mentioned, can you have a look?